### ISSUE TITLE
Hanging Lwt_main.run case

### ISSUE THREAD

**Issue by koen-struyve** [ocsigen/lwt#549]
The following code snippet

```
let test () =
  let waiter, wakener = Lwt.task () in
  let wrapped_t =
    begin
      let%lwt () = Lwt_unix.yield () in
      Lwt.return ()
    end
    [%finally
      let () = Lwt.wakeup wakener () in
      Lwt.return ()
    ]
  in
  let%lwt () = waiter in
  wrapped_t

let () = Lwt_main.run (test ())
```
hangs indefinitely, with wrapped_t remaining in a Sleep state. We noted that binding on the result of "test ()" first seems to resolve this.

This behavior is present from version 3.1.0 onwards.

**Comment by aantron** [comment]
Thanks for reporting. The attached commit fixes the issue. It is a bug in `bind`, `catch`, etc., that I accidentally introduced while reorganizing `lwt.ml` (#354). All the functions are fixed, and I also added test cases for all of them based on your snippet (thank you again).

I will look at couple more issues, and then make a release to make sure this fix reaches people soon.

**Comment by hcarty** [comment]
I see something similar with 3.2.1 - haven't tested against `master` yet to know if it's the same issue but the hanging effect is the same:
```ocaml
let l : unit Lwt.t list = []

let _, _ = Lwt_main.run (Lwt.nchoose_split [])
```
will hang indefinitely.  Tested in utop and compiled with ocamlopt, 64bit Linux w/libev.

**Comment by hcarty** [comment]
Looks like this is a separate issue that exists after `opam pin add lwt --dev-repo`.

**Comment by hcarty** [comment]
Issue opened in #551

**Comment by aantron** [comment]
@koen-struyve the fix is now out in [3.3.0](https://github.com/ocsigen/lwt/releases/tag/3.3.0). Thanks again for reporting.
